services/mirror/mirror_pull_safe_check.go

```diff
@@ -0,0 +1,196 @@
+					stderrMessage = sanitizer.Replace(stderr)
+					stdoutMessage = sanitizer.Replace(stdout)
+				}
+			}
+		}
+
+		// If there is still an error (or there always was an error)
+		if err != nil {
+			log.Error("SyncMirrors [repo: %-v]: failed to update mirror repository:\nStdout: %s\nStderr: %s\nErr: %v", m.Repo, stdoutMessage, stderrMessage, err)
+			desc := fmt.Sprintf("Failed to update mirror repository '%s': %s", newRepoPath, stderrMessage)
+			if err = admin_model.CreateRepositoryNotice(desc); err != nil {
+				log.Error("CreateRepositoryNotice: %v", err)
+			}
+			return false, nil
+		}
+	}
+	return false, nil
+}
+
+// detect user can update the mirror
+func detectCanUpdateMirror(ctx context.Context, m *repo_model.Mirror, gitArgs []string) (bool, error) {
+	repoPath := m.Repo.RepoPath()
+	newRepoPath := fmt.Sprintf("%s_update", repoPath)
+
+	// do copy directory recursive
+	err := util.CopyDir(repoPath, newRepoPath)
+	if err != nil {
+		log.Error("GetMirrorCanUpdate [repo: %-v]: CopyDirectory Error %v", m.Repo, err)
+		return false, err
+	}
+	syncStatus, err := syncRepoMirror(ctx, m, gitArgs, newRepoPath)
+	if err != nil {
+		return false, err
+	}
+	if !syncStatus {
+		return false, nil
+	}
+	gitCommitCountArgs := []string{"rev-list", "HEAD", "--count"}
+	stdoutNewRepoCommitCount, _, err := getGitCommandStdoutStderr(ctx, m, gitCommitCountArgs, newRepoPath)
+	if err != nil {
+		return false, err
+	}
+	stdoutNewRepoCommitCount = strings.TrimSpace(stdoutNewRepoCommitCount)
+	stdoutRepoCommitCount, _, err := getGitCommandStdoutStderr(ctx, m, gitCommitCountArgs, repoPath)
+	if err != nil {
+		return false, err
+	}
+	stdoutRepoCommitCount = strings.TrimSpace(stdoutRepoCommitCount)
+	var repoCommitCount, newRepoCommitCount int64
+	if i, err := strconv.ParseInt(stdoutRepoCommitCount, 10, 64); err == nil {
+		repoCommitCount = i
+	} else {
+		return false, err
+	}
+	if i, err := strconv.ParseInt(stdoutNewRepoCommitCount, 10, 64); err == nil {
+		newRepoCommitCount = i
+	} else {
+		return false, err
+	}
+	if repoCommitCount > newRepoCommitCount {
+		return false, nil
+	} else if repoCommitCount == newRepoCommitCount {
+		// noting to happen
+		return true, nil
+	} else {
+		// compare commit id
+		skipcout := newRepoCommitCount - repoCommitCount
+		gitNewRepoLastCommitIDArgs := []string{"log", "-1", fmt.Sprintf("--skip=%d", skipcout), "--format='%H'"}
+		stdoutNewRepoCommitID, _, err := getGitCommandStdoutStderr(ctx, m, gitNewRepoLastCommitIDArgs, newRepoPath)
+		if err != nil {
+			return false, err
+		}
+		gitRepoLastCommitIDArgs := []string{"log", "--format='%H'", "-n", "1"}
+		stdoutRepoCommitID, _, err := getGitCommandStdoutStderr(ctx, m, gitRepoLastCommitIDArgs, repoPath)
+		if err != nil {
+			return false, err
+		}
+		if stdoutNewRepoCommitID != stdoutRepoCommitID {
+			return false, fmt.Errorf("Old repo commit id: %s not match new repo id: %s", stdoutRepoCommitID, stdoutNewRepoCommitID)
+		}
+	}
+	return true, nil
```

Comment:
It is the purpose. However, I think copying the repo again and again is not ideal.

A better solution could be: track all branches locally, fetch remote, compare the merge-base, if the merge-base is not the locally tracked one, then there is a force-push.

And there might be other better solutions, but copying the repo is not the proper way.

Reply:
yep, I'm also looking for an ideal solution to figure out this sitution.

Reply:
Did you seem my option with the reference-transaction hook ? https://github.com/go-gitea/gitea/issues/14076#issuecomment-1000409708
You'll also need an option to allow/disallow force-push on specific branches otherwise mirrors will break quickly (people force-push on PR branches)